private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/maven/CoordinateGatheringListener.java

```diff
@@ -0,0 +1,91 @@
+
+import static com.github.bazelbuild.rules_jvm_external.resolver.maven.MavenPackagingMappings.mapPackagingToExtension;
+
+import com.github.bazelbuild.rules_jvm_external.Coordinates;
+import java.io.File;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+import org.eclipse.aether.AbstractRepositoryListener;
+import org.eclipse.aether.RepositoryEvent;
+import org.eclipse.aether.artifact.Artifact;
+import org.w3c.dom.Document;
+import org.xml.sax.SAXException;
+
+public class CoordinateGatheringListener extends AbstractRepositoryListener {
+
+  private final Map<Coordinates, Coordinates> knownRewrittenCoordinates = new ConcurrentHashMap<>();
+
+  @Override
+  public void artifactResolved(RepositoryEvent event) {
+
+    Artifact artifact = event.getArtifact();
+    if (!"pom".equals(artifact.getExtension())) {
+      return;
+    }
+
+    File file = event.getFile();
+    if (file == null) {
+      return;
+    }
+
+    // The correct thing to do here is to use a `ModelBuildingRequest` from Maven to build and
+    // construct a `Model`,
+    // which can then query the packaging of. However, since we just want one value from an XML
+    // document, we're
+    // going to do this the Old Skool way and just parse the document and run a little XPath over
+    // it.
+    try {
+      DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+      DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
+      Document doc = documentBuilder.parse(file);
+
+      doc.normalizeDocument();
+
+      XPath xPath = XPathFactory.newInstance().newXPath();
+      String packaging = xPath.compile("//project/packaging").evaluate(doc);
+
+      if (packaging == null) {
+        return;
+      }
+
+      packaging = packaging.trim();
+
+      String extension = mapPackagingToExtension(packaging);
+      // The default packaging is "jar" anyway
+      if (extension.isEmpty() || "jar".equals(extension)) {
+        return;
+      }
+
+      Coordinates coords =
+          new Coordinates(
+              artifact.getGroupId(), artifact.getArtifactId(), null, null, artifact.getVersion());
+
+      Coordinates actualCoords =
+          new Coordinates(
+              artifact.getGroupId(),
+              artifact.getArtifactId(),
+              extension,
+              artifact.getClassifier(),
+              artifact.getVersion());
+
+      knownRewrittenCoordinates.put(coords, actualCoords);
+    } catch (IOException
+        | ParserConfigurationException
+        | SAXException
+        | XPathExpressionException e) {
+      // Bail and hope for the best.
```

Comment:
Silently swallowing these seem error-prone, can we rethrow an unchecked exception?